File: MoInvoices/MoInvoices/ClientApp/src/app/services/invoice.service.ts
import { Injectable, Inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { DocumentType } from 'src/app/interfaces/documentType';
import { Invoice } from '../models/invoice.model';
import { InvoiceList } from '../interfaces/InvoiceList';
import { PaymentType } from '../interfaces/paymentType';
import { PaymentStatus } from '../interfaces/paymentStatus';

@Injectable()
export class InvoiceService {

  constructor(private httpClient: HttpClient, @Inject('BASE_URL') private baseUrl: string) { }

  createInvoice(invoice: Invoice) {
    invoice.userID = 1;
    return this.httpClient.post<Invoice>(this.baseUrl + 'invoice', invoice);
  }

  getInvoice(invoiceID: number) {
    return this.httpClient.get<Invoice>(this.baseUrl + `Invoice/${invoiceID}`);
  }

  deleteInvoice(invoiceID: number) {
    return this.httpClient.delete(this.baseUrl + `Invoice/${invoiceID}`);
  }

  updateInvoice(invoice: Invoice) {
    return this.httpClient.put(this.baseUrl + 'invoice', invoice);
  }

  getAllUserInvoices() {
    const userID = 1;
    return this.httpClient.get<InvoiceList[]>(this.baseUrl + `Invoice/AllUserInvoices/${userID}`);
  }

  // Metody uzupełniające Typy

  getDocumentTypes() {
    return this.httpClient.get<DocumentType[]>(this.baseUrl + `Invoice/DocumentTypes`);
  }

  getPaymentTypes() {
    return this.httpClient.get<PaymentType[]>(this.baseUrl + `Invoice/PaymentTypes`);
  }

  getPaymentStatuses() {
    return this.httpClient.get<PaymentStatus[]>(this.baseUrl + `Invoice/PaymentStatuses`);
  }
}

File: MoInvoices/MoInvoices.Core/Invoice/InvoiceService.cs
using AutoMapper;
using MoInvoices.Models;
using MoInvoices.Pages;
using System;
using System.Collections.Generic;
using System.Linq;
using MoInvoices.DTO;
using Microsoft.EntityFrameworkCore;
using MoInvoices.Data.Models;

namespace MoInvoices.Core
{
    public interface IInvoiceService
    {
        void AddNewInvoice(InvoiceDTO invoice);
        void UpdateInvoice(InvoiceDTO invoice);
        void DeleteInvoice(int id);
        InvoiceDTO GetInvoice(int invoiceID);
        IEnumerable<InvoiceListDTO> GetAllUserInvoices(int userID);


        IEnumerable<DocumentType> GetDocumentTypes();
        IEnumerable<PaymentType> GetPaymentTypes();
        IEnumerable<PaymentStatus> GetPaymentStatuses();
    }

    public class InvoiceService : IInvoiceService
    {
        private readonly MoInvoiceContext _context;
        private readonly IMapper _mapper;
        public InvoiceService(IMapper mapper, MoInvoiceContext context)
        {
            _context = context;
            _mapper = mapper;
        }

        public void AddNewInvoice(InvoiceDTO invoice)
        {
            try
            {
                var newInvoice = _mapper.Map<Invoice>(invoice);
                Contractor purchaser = _mapper.Map<Contractor>(invoice.Purchaser);
                Contractor vendor = _mapper.Map<Contractor>(invoice.Vendor);
                purchaser.ContractorTypeID = (int)Enum.ContractorType.Purchaser;
                vendor.ContractorTypeID = (int)Enum.ContractorType.Vendor;
    
                newInvoice.Contractors.Add(purchaser);
                newInvoice.Contractors.Add(vendor);

                _context.Invoice.Add(newInvoice);
                _context.SaveChanges();
            }
            catch (Exception ex)
            {
            }
        }

        public void UpdateInvoice(InvoiceDTO editedInvoice)
        {
            var invoice = _mapper.Map<Invoice>(editedInvoice);
            _context.Entry(invoice).State = EntityState.Modified;
            _context.SaveChanges();
        }

        public void DeleteInvoice(int id)
        {
            var invoice = _context.Invoice.Find(id);
            _context.Remove(invoice);

            _context.SaveChanges();
        }

        public InvoiceDTO GetInvoice(int invoiceID)
        {
            var invoice = _context.Invoice.Find(invoiceID);
            return _mapper.Map<InvoiceDTO>(invoice);
        }

        public IEnumerable<InvoiceListDTO> GetAllUserInvoices(int userID)
        {
            var invoiceList = _context.Invoice.Where(u => u.UserId.Equals(userID))
                                              .Select(x => _mapper.Map<InvoiceListDTO>(x));
       
            return invoiceList;
        }


        #region metody pobierające typy
        public IEnumerable<DocumentType> GetDocumentTypes()
        {
            return _context.DocumentType.ToList();
        }

        public IEnumerable<PaymentType> GetPaymentTypes()
        {
            return _context.PaymentType.ToList();
        }

        public IEnumerable<PaymentStatus> GetPaymentStatuses()
        {
            return _context.PaymentStatus.ToList();
        }
        #endregion

    }
}
